examples/medplum-provider/src/components/tasks/actions/AddDueDate.tsx

```diff
@@ -0,0 +1,101 @@
+import { Button, Modal } from '@mantine/core';
+import { useDisclosure } from '@mantine/hooks';
+import { notifications } from '@mantine/notifications';
+import { getQuestionnaireAnswers, MedplumClient, normalizeErrorString, PatchOperation } from '@medplum/core';
+import { Questionnaire, QuestionnaireResponse, Task, TaskRestriction } from '@medplum/fhirtypes';
+import { QuestionnaireForm, useMedplum } from '@medplum/react';
+import { IconCircleCheck, IconCircleOff } from '@tabler/icons-react';
+
+interface AddDueDateProps {
+  readonly task: Task;
+  readonly onChange: (updatedTask: Task) => void;
+}
+
+export function AddDueDate(props: AddDueDateProps): JSX.Element {
+  const medplum = useMedplum();
+  const [opened, { toggle, close }] = useDisclosure(false);
+
+  const handleAddDueDate = async (
+    date: string,
+    task: Task,
+    medplum: MedplumClient,
+    onChange: (task: Task) => void
+  ): Promise<void> => {
+    const taskId = task.id as string;
+
+    // We use a patch operation here to avoid race conditions. This ensures that if multiple users try to update the due-date simultaneously, only one will be successful.
+    const ops: PatchOperation[] = [{ op: 'test', path: '/meta/versionId', value: task.meta?.versionId }];
+
+    // Create a restriction period with the due-date if one doesn't already exist and add it to the PatchOperation. For more details on task due-dates, see https://www.medplum.com/docs/careplans/tasks#task-start--due-dates
+    const restriction: TaskRestriction = {
+      ...task.restriction,
+      period: {
+        ...task?.restriction?.period,
+        end: date,
+      },
+    };
+
+    const op: PatchOperation['op'] = task.restriction ? 'replace' : 'add';
+
+    ops.push({ op, path: '/restriction', value: restriction });
+
+    // Patch the resource with the new due-date
+    try {
+      const result = await medplum.patchResource('Task', taskId, ops);
+      notifications.show({
+        icon: <IconCircleCheck />,
+        title: 'Success',
+        message: 'Due-date updated.',
+      });
+      onChange(result);
+    } catch (error) {
+      notifications.show({
```

Comment:
Maybe a topic for discussion, but we use this pattern all over the place.  Probably overdue for a helper utility for "show success message" / "show error message".